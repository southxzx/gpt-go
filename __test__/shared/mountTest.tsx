import React from "react";
import { render } from "@testing-library/react";

export default function mountTest(Component: React.ComponentType) {
  describe("mount and unmount", () => {
    it("component could be updated and unmounted without errors", () => {
      const { rerender, unmount } = render(<Component />);
      expect(() => {
        rerender(<Component />);
        unmount();
      }).not.toThrow();
    });
  });
}
