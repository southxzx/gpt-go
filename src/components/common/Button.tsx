import React from "react";

import "../../ui.css";

type IButtonProps = React.ButtonHTMLAttributes<HTMLButtonElement>;

const Button: React.FC<IButtonProps> = ({ children, ...rest }) => {
  return (
    <button className="btn" {...rest}>
      {children}
    </button>
  );
};

export default Button;
