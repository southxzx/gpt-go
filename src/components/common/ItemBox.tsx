import React from "react";

import "../../ui.css";

interface IItemBoxProps {
  text: string;
  onClick: () => void;
}

const ItemBox: React.FC<IItemBoxProps> = ({ text, onClick }) => {
  return (
    <div className="item-box-wrapper" onClick={onClick}>
      <p>{text}</p>
    </div>
  );
};

export default ItemBox;
