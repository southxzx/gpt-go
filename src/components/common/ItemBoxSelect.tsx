import React, { useRef } from "react";
import IconArrowUp from "../../assets/IconArrowUp";
import useClickOutside from "../../hooks/useClickOutside";

import "../../ui.css";

interface IItemBoxProps {
  values: string[];
  placeholder?: string;
  onSelect: (value: string) => void;
}

const ItemBoxSelect: React.FC<IItemBoxProps> = ({
  values,
  onSelect,
  placeholder,
}) => {
  const [selectedValue, setSelectedValue] = React.useState<string>(
    placeholder || "..."
  );
  const [open, setOpen] = React.useState<boolean>(false);
  const contentRef = useRef<HTMLDivElement>(null);
  const selectRef = useRef<HTMLDivElement>(null);

  const handleOutsideClick = (e: MouseEvent) => {
    if (!selectRef.current?.contains(e.target as Node)) setOpen(false);
    e.stopPropagation();
  };

  useClickOutside(contentRef, handleOutsideClick);

  const onClick = () => {
    setOpen(!open);
  };

  const onSelectOption = (value: string) => {
    setSelectedValue(value);
    onSelect(value);
    setOpen(false);
  };

  return (
    <div className="dropdown-wrapper">
      <div
        className={`item-box-wrapper item-box-dropdown-wrapper ${
          open ? "active" : ""
        }`}
        onClick={onClick}
        ref={selectRef}
      >
        <p>{selectedValue}</p>
        <IconArrowUp />
      </div>
      {open && (
        <div className="dropdown-content" ref={contentRef}>
          {values.map((value) => (
            <div key={value} className="dropdown-option">
              <p onClick={() => onSelectOption(value)}>{value}</p>
            </div>
          ))}
        </div>
      )}
    </div>
  );
};

export default ItemBoxSelect;
