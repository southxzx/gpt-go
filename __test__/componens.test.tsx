import React from "react";
import { fireEvent, render } from "@testing-library/react";
import "@testing-library/jest-dom";

import Button from "../src/components/common/Button";
import ItemBox from "../src/components/common/ItemBox";
import ItemBoxSelect from "../src/components/common/ItemBoxSelect";

import mountTest from "./shared/mountTest";

describe("Button component", () => {
  mountTest(Button);
  mountTest(() => <Button disabled={true} />);
  mountTest(() => <Button loading={true} />);

  it("renders correctly", () => {
    const { container } = render(<Button />);
    expect(container.firstChild).toMatchSnapshot();
  });

  it("should not clickable when loading or disabled", () => {
    const onClick = jest.fn();
    const { container: containerLoading } = render(
      <Button loading={true} onClick={onClick} />
    );
    const { container: containerDisabled } = render(
      <Button disabled={true} onClick={onClick} />
    );
    fireEvent.click(containerLoading.firstChild as HTMLElement);
    expect(onClick).not.toHaveBeenCalledWith();

    fireEvent.click(containerDisabled.firstChild as HTMLElement);
    expect(onClick).not.toHaveBeenCalledWith();

    expect(containerDisabled.firstChild).toHaveClass("btn-disabled");
    expect(containerLoading.firstChild).toHaveClass("btn-disabled");
  });
});

describe("ItemBox component", () => {
  const onClick = jest.fn();
  mountTest(() => <ItemBox text="hello" onClick={onClick} />);

  it("renders correctly", () => {
    const { container } = render(<ItemBox text="hello" onClick={onClick} />);
    expect(container.firstChild).toMatchSnapshot();
  });

  it("should not clickable when disabled", () => {
    const { container } = render(
      <ItemBox text="hello" onClick={onClick} disabled={true} />
    );
    fireEvent.click(container.firstChild as HTMLElement);
    expect(onClick).not.toHaveBeenCalledWith();
    expect(container.firstChild).toHaveClass("disabled");
  });
});

describe("ItemBoxSelect component", () => {
  const onSelect = jest.fn();

  mountTest(() => (
    <ItemBoxSelect values={["1", "2", "3"]} onSelect={onSelect} />
  ));

  it("renders correctly", () => {
    const { container } = render(
      <ItemBoxSelect values={["1", "2", "3"]} onSelect={onSelect} />
    );
    expect(container.firstChild).toMatchSnapshot();
  });

  it("should not clickable when disabled", () => {
    const { container } = render(
      <ItemBoxSelect
        values={["1", "2", "3"]}
        onSelect={onSelect}
        disabled={true}
      />
    );
    fireEvent.click(container.firstChild as HTMLElement);
    expect(onSelect).not.toHaveBeenCalledWith();
    expect(container.firstChild?.firstChild).toHaveClass("disabled");
  });

  function toggleOpen(container: HTMLElement) {
    fireEvent.click(
      container.querySelector(".item-box-dropdown-wrapper") as HTMLElement
    );
  }

  it("should open and close when click", () => {
    const { container } = render(
      <ItemBoxSelect values={["1", "2", "3"]} onSelect={onSelect} />
    );
    toggleOpen(container);
    expect(container.firstChild?.firstChild).toHaveClass("active");
    expect(container.querySelector(".dropdown-content")).toBeInTheDocument();
    toggleOpen(container);
    expect(container.firstChild?.firstChild).not.toHaveClass("active");
    expect(
      container.querySelector(".dropdown-content")
    ).not.toBeInTheDocument();
  });

  it("should render options correctly", () => {
    const { container } = render(
      <ItemBoxSelect values={["1", "2", "3"]} onSelect={onSelect} />
    );
    toggleOpen(container);
    expect(container.querySelectorAll(".dropdown-option").length).toBe(3);
  });

  it('should choose option when click "1"', () => {
    const { container } = render(
      <ItemBoxSelect values={["1", "2", "3"]} onSelect={onSelect} />
    );
    toggleOpen(container);
    fireEvent.click(
      container.querySelector(".dropdown-option")?.firstChild as HTMLElement
    );
    expect(onSelect).toHaveBeenCalledWith("1");
  });

  it("should close when click outside", () => {
    const { container } = render(
      <ItemBoxSelect values={["1", "2", "3"]} onSelect={onSelect} />
    );
    toggleOpen(container);
    fireEvent.mouseDown(document);
    expect(container.firstChild?.firstChild).not.toHaveClass("active");
    expect(
      container.querySelector(".dropdown-content")
    ).not.toBeInTheDocument();
  });
});
