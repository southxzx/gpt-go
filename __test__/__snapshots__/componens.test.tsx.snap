// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Button component renders correctly 1`] = `
<button
  class="undefined btn "
/>
`;

exports[`ItemBox component renders correctly 1`] = `
<div
  class="item-box-wrapper "
>
  <p>
    hello
  </p>
</div>
`;

exports[`ItemBoxSelect component renders correctly 1`] = `
<div
  class="dropdown-wrapper"
>
  <div
    class="item-box-wrapper item-box-dropdown-wrapper  "
  >
    <p>
      ...
    </p>
    <svg
      fill="none"
      height="12"
      viewBox="0 0 12 12"
      width="12"
      xmlns="http://www.w3.org/2000/svg"
    >
      <g
        clip-path="url(#clip0_110_4423)"
      >
        <rect
          fill="white"
          fill-opacity="0.01"
          height="12"
          width="12"
        />
        <path
          d="M11.9045 3.20086L11.368 2.66433C11.2406 2.53698 11.0347 2.53698 10.9073 2.66433L6 7.5608L1.09269 2.66433C0.965338 2.53698 0.759399 2.53698 0.632042 2.66433L0.0955177 3.20086C-0.0318392 3.32822 -0.0318392 3.53416 0.0955177 3.66151L5.76967 9.33567C5.89703 9.46303 6.10297 9.46303 6.23033 9.33567L11.9045 3.66151C12.0318 3.53416 12.0318 3.32822 11.9045 3.20086Z"
          fill="#20133A"
        />
      </g>
      <defs>
        <clippath
          id="clip0_110_4423"
        >
          <rect
            fill="white"
            height="12"
            width="12"
          />
        </clippath>
      </defs>
    </svg>
  </div>
</div>
`;
